### PR TITLE
  Implement softmax with NNPACK

### PR DESCRIPTION
**Pull request by Knight-X**
Summary:
  Implement softmax calculation with nnpack package. This would
  accelerate computation in resource limited device.
  Add test case for softmax with NNPACK

### CODE REVIEW

**Comment by Yangqing** [comment]
@Maratyszcza would you mind taking a look?

File: caffe2/contrib/nnpack/nnpack_ops.cc [hunk 1]
```diff
@@ -335,9 +335,38 @@ class NNPACKLeakyReluOp final : public LeakyReluOp<float, CPUContext> {
...
+    // check if the machine has avx2 support.
+    OPERATOR_NEEDS_FEATURE(GetCpuId().avx2(), "NNPack requires AVX2");
+#endif
+  }
+
+  bool RunOnDevice() override {
+    auto& X = Input(0);
+    const auto N = X.dim32(0), C = X.dim32(1), H = X.dim32(2), W = X.dim32(3);
+    auto* Y = Output(0);
+    auto channels = C * H * W;
```
Comment (Maratyszcza) [inline caffe2/contrib/nnpack/nnpack_ops.cc]: This is actually not right - see https://github.com/caffe2/caffe2/blob/master/caffe2/operators/softmax_op.cc#L9-L13 for the right index calculation.

Comment (Knight-X) [reply caffe2/contrib/nnpack/nnpack_ops.cc]: Right, I will fix the error calculation.

**Comment by orionr** [comment]
Thank you for your contribution! PyTorch and Caffe2 are now officially merged with the PyTorch 1.0 preview release. Please submit on https://github.com/pytorch/pytorch if this is still an issue over there.
